coreax/approximation.py

```diff
@@ -292,3 +293,209 @@ def gramian_row_mean(self, x: Union[ArrayLike, Data], **kwargs) -> Array:
             self.num_train_points,
             self.base_kernel.gramian_row_mean,
         )
+
+
+class RegularisedInverseApproximator(ABC):
+    """
+    Base class for approximation methods to invert regularised kernel gram matrices.
+
+    When a dataset is very large, computing the regularised inverse of the kernel gram
+    matrix can be very time-consuming. Instead, this property can be approximated by
+    various methods. :class:`RegularisedInverseApproximator` is the base class
+    for implementing these approximation methods.
+
+    :param random_key: Key for random number generation
+    """
+
+    def __init__(self, random_key: KeyArrayLike):
+        """Define approximator to the kernel matrix inverse."""
+        # Assign inputs
+        self.random_key = random_key
+
+    @abstractmethod
+    def approximate(
+        self,
+        kernel_gramian: Array,
+        regularisation_parameter: float,
+        identity: Array,
+    ) -> Array:
+        r"""
+        Approximate regularised kernel matrix inverse.
+
+        .. note::
+            The function is designed to invert blocked "kernel matrices" where only the
+            top-left block contains non-zero elements. We return a block array, the same
+            size as the input array, where each block has only zero elements except
+            for the top-left block, which is the inverse of the non-zero input block.
+            The most efficient way to compute this in JAX requires the 'identity' array
+            to be a matrix of zeros except for ones on the diagonal up to the dimension
+            of the non-zero block.
+
+        :param kernel_gramian: Original :math:`n \times n` kernel gram matrix
+        :param regularisation_parameter: Regularisation parameter for stable inversion
+            of array, negative values will be converted to positive
+        :param identity: Block identity matrix
+        :return: Approximation of the kernel matrix inverse
+        """
+
+
+def randomised_eigendecomposition(
+    random_key: KeyArrayLike,
+    array: Array,
+    oversampling_parameter: int = 10,
```

Comment:
From the Halko paper, where `p` is the `oversampling_parameter`

"""
_Just how many extra samples do we need? Remarkably, for certain types of
random sampling schemes, the failure probability decreases superexponentially with
the oversampling parameter p. As a practical matter, setting p = 5 or p = 10
often gives superb results. This observation is one of the principal facts supporting
the randomized approach to numerical linear algebra._
"""

Also see:

"""
_The number of columns ℓ that the algorithm needs to reach this threshold is usually slightly larger than the minimal rank k of the smallest basis that verifies (4.2; **here 4.2. is the error the authors consider for the matrix that has a similar range to the target array**). We refer to this discrepancy p = ℓ − k as the oversampling parameter. The size of the oversampling parameter depends on several factors:

1. The matrix dimensions. Very large matrices may require more oversampling.
2. The singular spectrum. The more rapid the decay of the singular values, the less oversampling is needed. In the extreme case that the matrix has exact rank k, it is not necessary to oversample._ 
"""

Some experimentation led me to use larger values of `p` for the tests as the kernel matrix I used does not always have a rapidly decaying singular spectrum. The `oversampling_parameter ` should also not be any larger than the size of the matrix as then you are gaining nothing by this approach.

It could be worth implementing Algorithm 4.2 from the paper as this avoids leaving the choice of `oversampling_parameter` to the user.

On the `power_iterations` front Halko has this to say:
"""
Unfortunately, the simplest version of this scheme is inadequate in many
applications because the singular spectrum of the input matrix may decay slowly. To
address this difficulty, we incorporate q steps of a power iteration, where q = 1 or
q = 2 usually suffices in practice.
"""

Reply:
There is an element of diminishing returns, especially with `power_iterations`, it is one of the more costly parts of the algorithm and once you have reduced the error to machine precision it is pointless to keep beating the horse

Reply:
Thanks for the detail here - the default values seem sensible in this case and since the reference to the paper is there in the docstring, I'm happy to leave as is and the reader can refer to the paper if they want to change these.